File: LeetCodeProblemsGo/LeetCode440KthSmallestInLexicographicalOrder/LeetCode440KthSmallestInLexicographicalOrder.go
package LeetCode440KthSmallestInLexicographicalOrder

import "fmt"

func findKthNumber(n int, k int) int {
	current := 1
	steps := 0
	k--
	for k > 0 {
		steps = getNumCountInBetween(int64(current), int64(current+1), int64(n))
		if steps <= k {
			current++
			k -= steps
		} else {
			current *= 10
			k--
		}
	}
	return current
}
func getNumCountInBetween(first, second, n int64) int {
	result := 0
	var minValue int64
	for first <= n {
		minValue = second
		if n < minValue {
			minValue = n
		}
		result += int(minValue - first)
		first *= 10
		second *= 10
	}
	return result
}

func Calculate() {
	n := 17
	k := 2
	fmt.Println(findKthNumber(n, k))
}

File: LeetCodeProblemsCMake/Problems/LeetCode440KthSmallestInLexicographicalOrder.cpp
//
// Created by saeed on 6/12/25.
//

#include "LeetCode440KthSmallestInLexicographicalOrder.h"

#include <algorithm>
using namespace std;

int getNumsInBetween(long long first, long long second, long long n) {
    int result{};
    while (first <= n) {
        result+=min(second, n+1)-first;
        first*=10;
        second*=10;
    }
    return result;
}

int LeetCode440KthSmallestInLexicographicalOrder::findKthNumber(int n, int k) {
    int current{1};
    k--;
    int steps{};
    while (k>0) {
        steps=getNumsInBetween(current,current+1,n);
        if (steps<=k) {
            k-=steps;
            current++;
        }
        else {
            current*=10;
            k--;
        }
    }
    return current;
}
